### PR TITLE
[DPMMA-1443] Email report multisieries piechart

### PR DESCRIPTION
**Pull request by AlanWierzchonCA**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 

## Description

Adding a Pie chart to the raport with e-mail data: sent, read, clicked, unsubscribed.

image

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

2. Create and send several emails
3. Open and Read sent emails, click some links
4. Create a report with data source emails, and Graph Sent / Read / Clicked / Unsubscribed (Pie chart)
5. Check the results

### CODE REVIEW

**Comment by shinde-rahul** [comment]
[https://github.com/mautic/mautic/pull/14073](https://trello.com/c/r8jECh7U/3-https-githubcom-mautic-mautic-pull-14073)

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Mautic\CoreBundle\Helper\Chart;
```
Comment (shinde-rahul) [inline app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php:5]: Strict typing is missing.

```suggestion

declare(strict_types=1)

namespace Mautic\CoreBundle\Helper\Chart;
```

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+class SeriesPieChart extends AbstractChart implements ChartInterface
+{
+    /**
+     * Holds the suma of the all dataset values.
+     *
+     * @var float
+     */
+    protected $totalCount = 0;
```
Comment (shinde-rahul) [inline app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php]:
```suggestion
     */
    protected float $totalCount = 0;
```

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    public function setTotalCount(float|int $totalCount): void
+    {
+        $this->totalCount = $totalCount;
+    }
+
+    /**
+     * @param string[] $labels
+     */
+    public function setLabes(array $labels): void
```
Comment (shinde-rahul) [inline app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php]: Please fix the typo. `setLabes` should be `setLabels`

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                $data['backgroundColor'][]      = $color->toRgba(0.8);
+                $data['hoverBackgroundColor'][] = $color->toRgba(0.9);
+            }
+
+            $data['label'] = $this->labels[$datasetId];
+            $dataset[]     = $data;
+        }
+
+        return [
+            'labels'   => $this->labels,
```
Comment (shinde-rahul) [inline app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php:44]: To be on a safer side, 

```suggestion
            'labels'   => $this->labels ?? [],
```

**Review by shinde-rahul — CHANGES_REQUESTED** [review]
Thanks for the feature.

I have added few observations and suggestion.

**Comment by shinde-rahul** [comment]
@AlanWierzchonCA, if you are still working on the PR, please draft it. 

Thanks,
Rahul

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php (hunk 1 above)
Comment (AlanWierzchonCA) [reply app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php:5]: Ok, I added strict_types.

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php (hunk 4 above)
Comment (AlanWierzchonCA) [reply app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php:44]: I changed the declaration to 


`protected $labels = [];`

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php (hunk 3 above)
Comment (AlanWierzchonCA) [reply app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php]: Ok, I fixed it.

File: app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php (hunk 2 above)
Comment (AlanWierzchonCA) [reply app/bundles/CoreBundle/Helper/Chart/SeriesPieChart.php]: I changed to

`protected int|float $totalCount = 0;`

**Comment by AlanWierzchonCA** [comment]
Hi @shinde-rahul,

Thank you for your review.
I commited your suggestions and added SeriesPieChartTest for Unittest


Best,
Alan

File: app/bundles/CoreBundle/Tests/Helper/Chart/PieChartTest.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace Mautic\CoreBundle\Tests\Helper\Chart;
```
Comment (shinde-rahul) [inline app/bundles/CoreBundle/Tests/Helper/Chart/PieChartTest.php:5]:
```suggestion
declare(strict_types=1);

namespace Mautic\CoreBundle\Tests\Helper\Chart;
```

Comment (shinde-rahul) [reply app/bundles/CoreBundle/Tests/Helper/Chart/PieChartTest.php:5]: @AlanWierzchonCA, We should be adding `declare(strict_types=1);` always to the new class.

**Review by shinde-rahul — CHANGES_REQUESTED** [review]

File: app/bundles/EmailBundle/Entity/EmailRepository.php [hunk 6]
```diff
@@ -417,6 +426,107 @@ public function getEmailList($search = '', $limit = 10, $start = 0, $viewOther =
...
+            $results['not_read'] = $results['sent_count'] = $results['read_count'] = 0;
+        }
+
+        return $results;
+    }
+
+    /**
+     * @return int|array<string, int>
+     */
+    public function getUnsubscribedCount(QueryBuilder $queryBuilder): int|array
```
Comment (patrykgruszka) [inline app/bundles/EmailBundle/Entity/EmailRepository.php]: Is there any possibility that this method would return array? The `rowCount` should always return `int`

File: app/bundles/EmailBundle/Entity/EmailRepository.php [hunk 7]
```diff
@@ -417,6 +426,107 @@ public function getEmailList($search = '', $limit = 10, $start = 0, $viewOther =
...
+    {
+        $queryBuilder->resetQueryParts(['join']);
+        $this->addDNCTableForEmails($queryBuilder);
+        $queryBuilder->select('e.id as email_id, dnc.lead_id');
+        $queryBuilder->andWhere('dnc.reason='.DoNotContact::UNSUBSCRIBED);
+
+        return $queryBuilder->executeQuery()->rowCount();
+    }
+
+    public function getUniqueCliks(QueryBuilder $queryBuilder): int
```
Comment (patrykgruszka) [inline app/bundles/EmailBundle/Entity/EmailRepository.php]: Typo to fix: `getUniqueClicks`

**Review by patrykgruszka — CHANGES_REQUESTED** [review]
Nice feature! I've left a few suggestions on the code and encountered an issue while testing. Try to reproduce it:

1. Create a report and add the graphs.
2. Go to the report view and use the filter to select a date range that results in no emails (forcing an empty report). You should encounter the following error:

```
500 Internal Server Error - Mautic\EmailBundle\Entity\EmailRepository::getUniqueCliks(): Return value must be of type int, null returned 
```

File: app/bundles/EmailBundle/Entity/EmailRepository.php (hunk 7 above)
Comment (AlanWierzchonCA) [reply app/bundles/EmailBundle/Entity/EmailRepository.php]: I fixed it.

File: app/bundles/EmailBundle/Entity/EmailRepository.php (hunk 6 above)
Comment (AlanWierzchonCA) [reply app/bundles/EmailBundle/Entity/EmailRepository.php]: I fixed it.

**Comment by AlanWierzchonCA** [comment]
Thanks @patrykgruszka,  for your review. 
I sent the changes. 

The best 
Alana

**Review by shinde-rahul — APPROVED** [review]
LGTM, Thanks @AlanWierzchonCA!!!

**Comment by patrykgruszka** [comment]
Thanks for the fixes, it works great now! 

@rahuld-dev, did you test it or just do a code review?

**Review by patrykgruszka — APPROVED** [review]

**Review by andersonjeccel — APPROVED** [review]
Definitely nice!

**Pull request merged**